### PR TITLE
Polynomialization for ODEs

### LINKED ISSUE #3: Lift & Learn
https://arxiv.org/abs/1912.08177

### LINKED ISSUE #5: Operator Inference Model Order Reduction
https://arxiv.org/abs/2205.02304
https://arxiv.org/abs/2102.00083

### PR DESCRIPTION
**Pull request by bowenszhu**
preview of documentation https://bowenszhu.github.io/ModelOrderReduction.jl/previews/PR2/

$\frac{dx}{dt} = f(x)$

Linear projection based methods transform states $x$ to a subspace via $x = Vz$, but may still live in the $\mathcal O(n)$ space because of the nonlinear vector function $f$, where $n$ is the number of variables $x$ in the full-order system. 

$\frac{dz}{dt} = V^Tf(Vz)$

Some methods add additional approximation for the nonlinear function, such as DEIM. Some methods were designed to deal with the nonlinear function by, for example, exploiting some special forms of $f$ in order to pass $V^T$ or $V$ through $f(\cdot)$. Polynomial functions can be projected rather cheaply as $V^TB x \otimes x = V^TB (Vz) \otimes (Vz) = V^TB (V \otimes V)(z \otimes z)$. 

We aim to add a function, named `polynomialize` ~~or `quadratic_polynomialization`~~, to transfer system to an expanded system consisting of only ~~constant, linear and quadratic terms~~ *polynomials* without information loss. For example

```
julia> using ModelingToolkit

julia> @variables t x[1:3](t);

julia> D = Differential(t);

julia> eqs = [D(x[1]) ~ x[1] + x[1]^2 + 2exp(x[2]),
           D(x[2]) ~ x[2],
           D(x[3]) ~ x[1] + x[3] * x[2] + 3exp(x[2])];

julia> @named sys = ODESystem(eqs, t, x, []; checks = false);

julia> new_sys = polynomialize(sys);

julia> states(new_sys)
4-element Vector{Term{Real, Base.ImmutableDict{DataType, Any}}}:
 x[1](t)
 x[2](t)
 x[3](t)
 y₁(t)

julia> equations(new_sys)
4-element Vector{Equation}:
 Differential(t)(x[1](t)) ~ 2y₁(t) + x[1](t)^2 + x[1](t)
 Differential(t)(x[2](t)) ~ x[2](t)
 Differential(t)(x[3](t)) ~ 3y₁(t) + x[2](t)*x[3](t) + x[1](t)
 Differential(t)(y₁(t)) ~ y₁(t)*x[2](t)
```

Furthermore, polynomialization facilitates methods of lift & learn, operator inference, more accurate computation of quadrature.

resolve #3

### CODE REVIEW

**Comment by FHoltorf** [comment]
Very good work, this is going in a really nice direction. 

Just one thought: Why restrict to linear-quadratic models? It has been a while since I read the Wilcox group papers on that but my understanding was that this choice was mostly motivated because it is the next best thing that is not linear and still "easy" (bookkeeping-wise). Of course any polynomial dynamical system can be lifted to quadratic form but a) it is not clear to me that that is better for performance than just projecting the polynomial model and b) there are differential-algebraic polynomial systems that cannot be lifted to quadratic form (IIRC one needs to go to cubic models in that case)!

**Comment by FHoltorf** [comment]
maybe @ChrisRackauckas has thoughts on this too

**Comment by bowenszhu** [comment]
> Just one thought: Why restrict to linear-quadratic models? It has been a while since I read the Willcox group papers on that but my understanding was that this choice was mostly motivated because it is the next best thing that is not linear and still "easy" (bookkeeping-wise). Of course any polynomial dynamical system can be lifted to quadratic form but a) it is not clear to me that that is better for performance than just projecting the polynomial model and b) there are differential-algebraic polynomial systems that cannot be lifted to quadratic form (IIRC one needs to go to cubic models in that case)!
> 
> It seems to me that the hardest part about this is not to find the projection (for which case linear-quadratic would be easier than polynomial) but rather to find the immersion that maps the nonlinear system to a linear-quadratic one. It seems finding an immersion that maps to arbitrary polynomials could be easier to automate. Thoughts? Because if so, we might want to consider going for the general polynomial case instead.

True. Going for general polynomials is more practical. I will go this way.

Initially I wanted to make a quadratic-linear polynomialization because the papers of lift & learn and operator inference demonstrate the quadratic case $Ax+H(x\otimes x)$. But I found that if a high-degree monomial exists in the model, it requires a very large amount of new variables and differential and algebraic equations.

**Comment by bowenszhu** [comment]
As of commit [f4456c9](https://github.com/SciML/ModelOrderReduction.jl/pull/22/commits/f4456c9e33ef05b9e3847ea09061994a0f6b98fd), `polynomialization` is capable of something as follows.

```
@variables t x(t)
D = Differential(t)
eqs = [D(x) ~ exp(x)]
@named sys = ODESystem(eqs, t, [x], []; checks = false)
new_sys = polynomialization(sys)
ModelingToolkit.get_eqs(new_sys)

2-element Vector{Equation}:
 Differential(t)(x(t)) ~ y₁(t)
 Differential(t)(y₁(t)) ~ y₁(t)^2
```

```
@variables t x(t)
D = Differential(t)
eqs = [D(x) ~ sin(x)]
@named sys = ODESystem(eqs, t, [x], []; checks = false)
new_sys = polynomialization(sys)
ModelingToolkit.get_eqs(new_sys)

3-element Vector{Equation}:
 Differential(t)(x(t)) ~ y₁(t)
 Differential(t)(y₁(t)) ~ y₁(t)*y₂(t)
 Differential(t)(y₂(t)) ~ -(y₁(t)^2)
```

```
@variables t x[1:3](t)
D = Differential(t)
eqs = [D(x[1]) ~ x[1] + x[3]^4 + 7,
    D(x[2]) ~ x[2] + 3sin(x[3])^3,
    D(x[3]) ~ x[1] + 5cos(exp(x[2])) + 4exp(x[1])]
@named sys = ODESystem(eqs, t, x, []; checks = false)
new_sys = polynomialization(sys)
ModelingToolkit.get_eqs(new_sys)

9-element Vector{Equation}:
 Differential(t)(x[1](t)) ~ 7 + x[3](t)^4 + x[1](t)
 Differential(t)(x[2](t)) ~ 3(y₁(t)^3) + x[2](t)
 Differential(t)(x[3](t)) ~ 5y₃(t) + 4y₆(t) + x[1](t)
 Differential(t)(y₁(t)) ~ (5y₃(t) + 4y₆(t) + x[1](t))*y₂(t)
 Differential(t)(y₂(t)) ~ -(5y₃(t) + 4y₆(t) + x[1](t))*y₁(t)
 Differential(t)(y₃(t)) ~ -(3(y₁(t)^3) + x[2](t))*y₄(t)*y₅(t)
 Differential(t)(y₄(t)) ~ (3(y₁(t)^3) + x[2](t))*y₄(t)
 Differential(t)(y₅(t)) ~ (3(y₁(t)^3) + x[2](t))*y₃(t)*y₄(t)
 Differential(t)(y₆(t)) ~ (7 + x[3](t)^4 + x[1](t))*y₆(t)
```

**Comment by bowenszhu** [comment]
As of commit [900b810](https://github.com/SciML/ModelOrderReduction.jl/pull/22/commits/900b81042548b8c41ba295867e12bb4df4e0a0ee), `polynomialization` is capable of something as follows.

```
@variables t x(t)
D = Differential(t)
eqs = [D(x) ~ -exp(x)]
@named sys = ODESystem(eqs, t, [x], []; checks = false)
new_sys = polynomialization(sys)
ModelingToolkit.get_eqs(new_sys)

2-element Vector{Equation}:
 Differential(t)(x(t)) ~ -y₁(t)
 Differential(t)(y₁(t)) ~ -(y₁(t)^2)
```

```
@variables t x(t)
D = Differential(t)
eqs = [D(x) ~ sin(x)]
@named sys = ODESystem(eqs, t, [x], []; checks = false)
new_sys = polynomialization(sys)
ModelingToolkit.get_eqs(new_sys)

3-element Vector{Equation}:
 Differential(t)(x(t)) ~ y₁(t)
 Differential(t)(y₁(t)) ~ y₁(t)*y₂(t)
 Differential(t)(y₂(t)) ~ -(y₁(t)^2)
```

```
@variables t x[1:3](t)
D = Differential(t)
eqs = [D(x[1]) ~ x[1] + x[3]^4 + 7,
    D(x[2]) ~ x[2] + 3sin(x[3])^3,
    D(x[3]) ~ x[1] + 5cos(exp(x[2])) + 4exp(x[1])]
@named sys = ODESystem(eqs, t, x, []; checks = false)
new_sys = polynomialization(sys)
ModelingToolkit.get_eqs(new_sys)

9-element Vector{Equation}:
 Differential(t)(x[1](t)) ~ 7 + x[3](t)^4 + x[1](t)
 Differential(t)(x[2](t)) ~ 3(y₁(t)^3) + x[2](t)
 Differential(t)(x[3](t)) ~ 5y₄(t) + 4y₆(t) + x[1](t)
 Differential(t)(y₁(t)) ~ (5y₄(t) + 4y₆(t) + x[1](t))*y₂(t)
 Differential(t)(y₂(t)) ~ -(5y₄(t) + 4y₆(t) + x[1](t))*y₁(t)
 Differential(t)(y₃(t)) ~ (3(y₁(t)^3) + x[2](t))*y₃(t)
 Differential(t)(y₄(t)) ~ -(3(y₁(t)^3) + x[2](t))*y₃(t)*y₅(t)
 Differential(t)(y₅(t)) ~ (3(y₁(t)^3) + x[2](t))*y₃(t)*y₄(t)
 Differential(t)(y₆(t)) ~ (7 + x[3](t)^4 + x[1](t))*y₆(t)
```

**Comment by bowenszhu** [comment]
WIP: changing to use term rewriter

File: src/lift_learn.jl [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        add_dict = Dict()
+        sizehint!(add_dict, length(add.dict))
+        for (term, coeff) in add.dict
+            var = change_variables(term)
+            add_dict[var] = coeff
+        end
+        return SymbolicUtils.Add(Real, add.coeff, add_dict)
+    end
+
+    function change_variables(mul::SymbolicUtils.Mul)::SymbolicUtils.Symbolic
```
Comment (bowenszhu) [inline src/lift_learn.jl:48]: Change to `isadd`, `ismul`, etc.

**Pull request closed without merging**